### ISSUE TITLE
geti

### ISSUE THREAD

**Issue by YimSungGoo** [YimSungGoo/daily#37]
https://github.com/open-edge-platform/geti
Intel® Arc™ B 시리즈 GPU에서 모델 미세 조정을 실행하여 AI 구축을 시작할 수 있습니다. 컴퓨터 비전 솔루션을 배포할 준비가 되면 전체 Intel XPU 포트폴리오에서 OpenVINO™에 최적화된 모델을 쉽게 실행할 수 있습니다.

## ● YOLO 모델 호환성 해결 방안:

  🚨 현재 상황 분석

  - 사용 중: YOLOv8n, YOLO12n
  - Geti 지원: YOLOX만 공식 지원
  - 문제: 직접적인 모델 호환성 부족

  💡 해결 방안들

  🔄 방안 1: 모델 변환/이식

  YOLO → YOLOX 변환:
    - 가중치 구조 매핑
    - 아키텍처 차이점 보정
    - 성능 검증 필요

  장점: 기존 학습 데이터 활용
  단점: 변환 과정 복잡, 성능 저하 가능

  🆕 방안 2: Geti에서 새로 학습

  신규 학습 프로세스:
    1. 기존 데이터셋 Geti로 이전
    2. YOLOX 모델로 재학습
    3. 성능 비교 & 튜닝

  장점: Geti 최적화, 안정성
  단점: 처음부터 다시 학습

  🔧 방안 3: 하이브리드 접근

  병행 전략:
    - 현재: YOLOv8n/YOLO12n 계속 사용
    - 신규: Geti + YOLOX로 실험
    - 비교: 성능/효율성 평가 후 결정

  📈 방안 4: 커스텀 통합

  기술적 해결:
    - OpenVINO 변환: YOLO → OpenVINO IR
    - Geti SDK 활용: 외부 모델 통합
    - 파이프라인: 전처리/후처리 통일

  🛠️ 구체적 실행 방안

  단기 (1-2주)

  1. 성능 벤치마크
  # 현재 모델 성능 측정
  YOLOv8n vs YOLOv12n vs YOLOX
  - mAP, 속도, 메모리 사용량
  2. Geti YOLOX 테스트
  테스트 계획:
  - 소규모 데이터셋으로 YOLOX 학습
  - 기존 YOLO 모델과 성능 비교
  - 학습 시간/리소스 비교

  중기 (1-2개월)

  3. 데이터 이전 & 재학습
  이전 프로세스:
  - 어노테이션 형식 변환 (YOLO → Geti)
  - 클래스 매핑 확인
  - 검증 데이터셋 분할
  4. 성능 최적화
  튜닝 영역:
  - 하이퍼파라미터 조정
  - 데이터 증강 설정
  - Active Learning 활용

##   🎯 핵심 해결책

  Training Extensions는 모듈형 아키텍처로 설계되어 있어 새로운 모델 추가가 가능합니다.

  🛠️ 모델 추가 방법

  1단계: 모델 아키텍처 구현
```
  # src/otx/models/detection/yolov8.py
  from otx.models.base import BaseModel

  class YOLOv8Model(BaseModel):
      def __init__(self, config):
          super().__init__(config)
          # YOLOv8 아키텍처 구현

      def forward(self, x):
          # YOLOv8 forward pass
          pass

      def loss_function(self, pred, target):
          # YOLOv8 loss 구현
          pass
```
  2단계: 모델 레시피 생성
```
  # recipes/detection/yolov8n.yaml
  model:
    name: "yolov8n"
    backbone: "yolov8_backbone"
    neck: "yolov8_neck"
    head: "yolov8_head"

  training:
    optimizer: "SGD"
    lr: 0.01
    batch_size: 16
    epochs: 100

  data:
    input_size: [640, 640]
    mean: [0.485, 0.456, 0.406]
    std: [0.229, 0.224, 0.225]
```
  3단계: 모델 팩토리 등록
```
  # src/otx/models/__init__.py
  from .detection.yolov8 import YOLOv8Model

  MODEL_REGISTRY = {
      "yolov8n": YOLOv8Model,
      "yolo12n": YOLO12Model,  # 추가
      # 기존 모델들...
  }
```
  🔧 구체적 구현 방법

  방법 1: 직접 구현 (권장)
```
  # 1. Training Extensions 소스 클론
  git clone https://github.com/open-edge-platform/training_extensions
  cd training_extensions

  # 2. 개발 환경 설정
  pip install -e .

  # 3. 새 모델 브랜치 생성
  git checkout -b add-yolov8-support

  # 모델 구현 예제
  # src/otx/algorithms/detection/adapters/openvino/yolov8.py

  class YOLOv8OpenVINOAdapter:
      def __init__(self, model_path):
          self.model = self.load_model(model_path)

      def preprocess(self, image):
          # YOLOv8 전처리
          return processed_image

      def postprocess(self, outputs):
          # YOLOv8 후처리 (NMS 등)
          return detections
```
  방법 2: YOLOv8 래퍼 활용
```
  # YOLOv8 공식 모델을 Training Extensions에 통합
  from ultralytics import YOLO

  class YOLOv8Wrapper(BaseModel):
      def __init__(self, config):
          self.yolo_model = YOLO(config.model_path)

      def train(self, dataset):
          # YOLOv8 학습을 Training Extensions API로 래핑
          results = self.yolo_model.train(
              data=dataset,
              epochs=self.config.epochs
          )
          return results
```
  방법 3: 모델 변환 및 통합
```
  # 기존 YOLOv8 가중치를 Training Extensions 형식으로 변환
  def convert_yolov8_weights(yolov8_path, output_path):
      # 가중치 구조 매핑
      yolo_weights = torch.load(yolov8_path)
      otx_weights = {}

      # 레이어별 매핑
      for key, value in yolo_weights.items():
          otx_key = map_layer_name(key)  # 키 변환
          otx_weights[otx_key] = value

      torch.save(otx_weights, output_path)
```
  📋 필요한 작업들

  핵심 구현 요소

  1. 모델 아키텍처: YOLOv8/YOLO12 네트워크 구조
  2. 손실 함수: YOLOv8 특화 loss function
  3. 데이터 로더: YOLOv8 입력 형식 지원
  4. 후처리: NMS, confidence filtering
  5. 메트릭: mAP, precision, recall 계산
  6. OpenVINO 변환: IR 형식 지원

  설정 파일 확장
```
  # configs/detection/yolov8_variants.yaml
  models:
    yolov8n:
      backbone: "yolov8n_backbone"
      parameters: 3.2M

    yolov8s:
      backbone: "yolov8s_backbone"
      parameters: 11.2M

    yolo12n:
      backbone: "yolo12n_backbone"
      parameters: 2.8M
```
  🚀 실행 계획

  1단계: 프로토타입 (1-2주)
```
  # 최소 기능 구현
  - YOLOv8n 기본 학습 지원
  - 단순 추론 기능
  - 기본 메트릭 계산
```
  2단계: 완전 통합 (2-4주)
```
  # 전체 기능 구현
  - 모든 YOLOv8 변형 지원
  - Active Learning 연동
  - OpenVINO 최적화
  - Geti UI 통합
```
  3단계: YOLO12 확장 (2-3주)
```
  # YOLO12 추가
  - YOLO12 아키텍처 구현
  - 성능 벤치마크
  - 문서화 및 테스트
```
  💡 추천 접근법

  단기 해결책

  1. YOLOv8 래퍼 방식: 빠른 프로토타입
  2. 기존 모델 활용: YOLOX 기반 수정

  장기 해결책

  1. 완전 통합: 네이티브 지원
  2. 커뮤니티 기여: 오픈소스 PR 제출

  🎯 결론: Training Extensions의 모듈형 구조를 활용하면 YOLOv8/YOLO12 지원이 충분히 가능합니다. 가장 효율적인 방법은 YOLOv8 래퍼를 먼저 구현하고, 
  점진적으로 네이티브 통합으로 발전시키는 것입니다.

### Support custom model architectures? (github.com/open-edge-platform/geti/issues/187)

> armham01
> Is there a way to include custom models or e.g yolo 11, etc? Or do I have zo stick to what is selectable in the platform?

> olkham
> We use https://github.com/open-edge-platform/training_extensions for the training backend. So contributing a model there would see it included when building the platform - however, I'll be honest, it's not trivial today. We'll work on improving the experience and document the process.

<img width="811" height="1292" alt="Image" src="https://github.com/user-attachments/assets/9ce1d324-a3c9-4d54-809c-f4d0d92818b4" />

● YOLOX vs YOLOv8 상세 비교 분석:

  🎯 핵심 차이점 요약

  Head (검출 헤드) 구조

  YOLOX Head

  구조:
  - Decoupled Head (분리형 헤드)
  - 1×1 Conv → 채널 수를 256으로 축소
  - 2개 병렬 브랜치:
    ├── Classification: 2개의 3×3 Conv (클래스 예측)
    └── Regression: 2개의 3×3 Conv + IoU 브랜치 (위치 예측)

  YOLOv8 Head

  구조:
  - Anchor-Free Split Ultralytics Head
  - TaskAlignedAssigner 기반 동적 할당
  - 향상된 Decoupled 구조:
    ├── Classification Branch (분류)
    └── Detection Branch (위치 + IoU)
  - Objectness 브랜치 제거

  🔍 차이점: YOLOv8이 더 정교한 동적 할당과 효율적인 구조

  ---
  Neck (특징 융합) 구조

  YOLOX Neck

  구조:
  - 표준 FPN (Feature Pyramid Network)
  - 단순한 top-down + lateral connection
  - 기본적인 다중 스케일 특징 융합

  YOLOv8 Neck

  구조:
  - PANet (Path Aggregation Network) 기반
  - C2f 모듈 활용:
    ├── 고차원 의미 정보 + 저차원 공간 정보 결합
    ├── 더 많은 skip connection
    └── 향상된 정보 흐름
  - 공간 주의 메커니즘 통합

  🔍 차이점: YOLOv8이 훨씬 정교한 특징 융합과 정보 흐름

  ---
  Loss Function (손실 함수)

  YOLOX Loss

  구성:
  - SimOTA (Simplified Optimal Transport Assignment)
  - Focal Loss (분류)
  - IoU Loss (회귀)
  - 전통적인 라벨 할당 전략

  YOLOv8 Loss

  구성:
  - TaskAlignedAssigner (TOOD 기반)
  - Distribution Focal Loss (회귀)
  - 동적 라벨 할당:
    └── 분류 점수 × IoU 가중합 기반 선택
  - 향상된 정렬 메트릭

  🔍 차이점: YOLOv8이 더 스마트한 동적 할당과 정교한 손실 함수

  ---
  📊 성능 차이점

  정확도 & 속도

  YOLOv8 우위:
  - 동일 모델 크기 대비 더 높은 mAP
  - 더 적은 파라미터로 동등/우수 성능
  - 더 빠른 추론 속도
  - 향상된 작은 객체 탐지

  기능 범위

  YOLOX: 객체 탐지 전용
  YOLOv8: 멀티태스크
  ├── Object Detection
  ├── Instance Segmentation
  ├── Pose Estimation
  ├── Image Classification
  └── Object Tracking

  ---
  ⚖️ 호환성 분석

  🟡 중간 정도 차이

  유사점:
  ✅ 둘 다 Anchor-Free
  ✅ 둘 다 Decoupled Head 사용
  ✅ 기본 백본 구조 유사

  차이점:
  ❌ Neck 구조 상당히 다름 (FPN vs PANet+C2f)
  ❌ Loss 함수 완전히 다름 (SimOTA vs TaskAligned)
  ❌ Head 세부 구현 차이

  🔧 변경 필요 수준

  변경 범위:
  1. Neck: C2f 모듈 구현 필요 (상당한 작업)
  2. Head: TaskAlignedAssigner 적용 (중간 작업)
  3. Loss: Distribution Focal Loss 구현 (중간 작업)
  4. 데이터 파이프라인: 호환성 확인 필요

  ---
  💡 실용적 권장사항

  단기 해결책 (2-4주)

  우선순위:
  1. YOLOX 기반으로 시작 (기존 지원)
  2. YOLOv8 Head만 부분 적용
  3. 성능 비교 후 점진적 확장

  장기 해결책 (2-3개월)

  완전 마이그레이션:
  1. C2f 모듈 구현
  2. TaskAlignedAssigner 적용
  3. Distribution Focal Loss 통합
  4. 전체 YOLOv8 파이프라인 구축

  🎯 결론: YOLOX와 YOLOv8은 기본 철학은 비슷하지만, Neck과 Loss에서 상당한 차이가 있습니다. 완전한 YOLOv8 지원을 위해서는 중간~상당한 수준의 개발 작업이 
  필요합니다.

**Comment by YimSungGoo** [comment]
https://open-edge-platform.github.io/training_extensions/latest/guide/get_started/installation.html

https://open-edge-platform.github.io/training_extensions/latest/guide/tutorials/base/how_to_train/detection.html

Dataset preparation

[COCO](https://cocodataset.org/#format-data)

[Pascal-VOC](https://openvinotoolkit.github.io/datumaro/stable/docs/data-formats/formats/pascal_voc.html)

[YOLO](https://openvinotoolkit.github.io/datumaro/stable/docs/data-formats/formats/yolo.html)

Export
1. otx export exports a trained Pytorch .pth model to the OpenVINO™ Intermediate Representation (IR) format. It allows to efficiently run it on Intel hardware, especially on CPU, using OpenVINO™ runtime. Also, the resulting IR model is required to run PTQ optimization in the section below. IR model contains 2 files: exported_model.xml for architecture and exported_model.bin for weights.


https://open-edge-platform.github.io/training_extensions/latest/guide/get_started/cli_commands.html
